servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HttpRequestAutoDrainingServiceFilter.java

```diff
@@ -0,0 +1,111 @@
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+import io.servicetalk.http.utils.BeforeFinallyHttpOperator;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static io.servicetalk.concurrent.api.Single.defer;
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+public final class HttpRequestAutoDrainingServiceFilter implements StreamingHttpServiceFilterFactory {
+
+    static final StreamingHttpServiceFilterFactory INSTANCE = new HttpRequestAutoDrainingServiceFilter();
+
+    private HttpRequestAutoDrainingServiceFilter() {
+        // singleton
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            return defer(() -> doHandle(ctx, request, responseFactory).shareContextOnSubscribe());
+        }
+
+        private Single<StreamingHttpResponse> doHandle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            final AtomicBoolean payloadSubscribed = new AtomicBoolean();
+            request.transformMessageBody(body ->
+                    body.beforeSubscriber(() -> {
+                        payloadSubscribed.set(true);
+                        return HttpMessageDiscardWatchdogServiceFilter.NoopSubscriber.INSTANCE;
+                    }));
+            return delegate().handle(ctx, request, responseFactory)
+                    .liftSync(new BeforeFinallyHttpOperator(new TerminalSignalConsumer() {
+                @Override
+                public void onComplete() {
+                    if (payloadSubscribed.compareAndSet(false, true)) {
+                        request.payloadBody().ignoreElements().shareContextOnSubscribe().subscribe();
+                    }
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+                    cancel(); // same behavior as cancel()
+                }
+
+                @Override
+                public void cancel() {
+                    maybeCancelMessageBody(payloadSubscribed, request.messageBody());
```

Comment:
Now that we use a single operator, we can implement all logic here instead of adding 2 more methods. Might be easier to see all state changes of `payloadSubscribed` if they have same indentation